File: 15-3sum/3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        int n = nums.size();
        vector<int> sorted = nums;
        vector<vector<int>> res;
        sort(sorted.begin(), sorted.end());

        for(int k = 0; k < n; k++) {
            if(k > 0 && sorted[k] == sorted[k-1]) {
                continue;
            }
            int i = k+1, j = n-1;
            while(i<j) {
                int sum = sorted[i] + sorted[j] + sorted[k];
                if (sum < 0) {
                    i++;
                } else if (sum > 0) {
                    j--;
                } else {
                    res.push_back({sorted[i], sorted[j], sorted[k]});
                    i++;
                    while(sorted[i] == sorted[i-1] & i<j){
                        i++;
                    }
                }
            }
        }

        return res;
    }
};

File: 15-3sum/3sum.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        res = []
        k = 0

        numSorted = sorted(nums)

        for k in range(len(numSorted)):
            if k > 0 and numSorted[k] == numSorted[k-1]:
                continue
            
            i, j = k + 1, len(numSorted) - 1

            while i < j:
                s = numSorted[i] + numSorted[j] + numSorted[k]
                if s > 0:
                    j -= 1
                elif s < 0:
                    i += 1
                else:
                    res.append([numSorted[i], numSorted[j], numSorted[k]])
                    i += 1
                    while numSorted[i] == numSorted[i-1] and i < j:
                        i += 1
        
        return res
